data/transactions/logic/eval.go

```diff
@@ -218,22 +224,28 @@ func (ep EvalParams) log() logging.Logger {
 type evalContext struct {
 	EvalParams
 
-	stack   []stackValue
-	program []byte // txn.Lsig.Logic ?
-	pc      int
-	nextpc  int
-	err     error
-	intc    []uint64
-	bytec   [][]byte
-	version uint64
-	scratch [256]stackValue
-
-	stepCount int
-	cost      int
-
-	// Ordered set of pc values that a branch could go to.
-	// If Check pc skips a target, the source branch was invalid!
-	branchTargets []int
+	stack     []stackValue
+	callstack []int
+	program   []byte // txn.Lsig.Logic ?
+	pc        int
+	nextpc    int
+	err       error
+	intc      []uint64
+	bytec     [][]byte
+	version   uint64
+	scratch   [256]stackValue
+
+	cost int // cost incurred so far
```

Comment:
what is the point of having `cost` as a member of context? I guess the less shared vars the better (imo returning a pair (cost, err) from checkers is fine)

Reply:
I agree, I can probably get rid of it.  This is the "old" static kind of cost, it's not new code with this PR.  go just reformatted the line because I got rid of stepCount, above it.

Reply:
I take that back. It could have been removed, if we only cared about the old static check. It would be reasonable for checkStep to simply return the cost, as you suggest, rather than modify state.  The program checker would sum them, and if it's too high, reject the program.

But when doing dynamic checking, you want to stop as soon as you decode the instruction that would exceed the budget.  That means the total cost so far is truly part of the execution environment.  That final instruction should never even run.  So you would have to pass the running total into the step() method.  But that's exactly what the evalContext should represent.

That is, this code in `step()` needs to work as it does:
```
	cx.cost += deets.Cost
	if cx.cost > cx.budget() {
		cx.err = fmt.Errorf("%3d %s budget exceeded", cx.pc, spec.Name)
		return
	}
	spec.op(cx)
```